### PR TITLE
Manual Audience Segments API write endpoints

### PR DESCRIPTION
**Pull request by filleduchaos**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

API endpoints (`create`, `destroy`, `add_users`, `remove_users`) that modify audience segments.

This is still a fairly large PR, but a good deal of it is Swagger docs

## Related Tickets & Documents

- Related Issue #19438 

## QA Instructions, Screenshots, Recordings

These endpoints should all only work with an API key that has admin access to `AudienceSegment`, and should fail if used with a non-manual audience segment ID (returning 404)

- `POST /api/segments` should create a manual audience segment and return it.
- `DELETE /api/segments/{id}` should delete the provided manual segment and return it
    - It should return a readable error if the segment is currently being used by a billboard
    - You can add a segment to a billboard with `DisplayAdd.find(billboard_id).update!(audience_segment: segment)`
- `PUT /api/segments/{id}/add_users` should add the users with the IDs passed in the request body (`{ user_ids: [<ids>] }`) to the provided manual segment.
    - It should return JSON containing `"succeeded` (an array of user IDs that were processed successfully and `"failed"` (user IDs that could not be added to the segment)
    - It should return a readable error if user IDs are not provided or if more than 10,000 are provided
- `PUT /api/segments/{id}/remove_users` should work the same as above, but removing users from the segment instead of adding them.

## Added/updated tests?

- [x] Yes

## [optional] Are there any post deployment tasks we need to perform?

I am not quite sure - there's a migration that introduces a new index, which might need to be tested on canary first

### CODE REVIEW

File: app/services/bulk_segmented_users.rb [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        created_at: now,
+        updated_at: now
+      }
+    end
+
+    upserted_user_ids = SegmentedUser.upsert_all(
+      segmented_users,
+      unique_by: :index_segmented_users_on_audience_segment_and_user,
+      update_only: [:updated_at],
+      record_timestamps: false,
```
Comment (jaw6) [inline app/services/bulk_segmented_users.rb:25]: Minor: I found this surprising — seeing as we're setting `created_at` and `updated_at` explicitly on 16/17, I assume there's a reason we're turning `timestamps` off here?

File: app/services/bulk_segmented_users.rb [hunk 2]
```diff
@@ -0,0 +1,51 @@
+class BulkSegmentedUsers
```
Comment (jaw6) [inline app/services/bulk_segmented_users.rb:1]: I bring this up reluctantly, but we have [a few conventions we try to follow with service classes](https://developers.forem.com/backend/service-objects#what-are-service-objects) — `app/services/{things}/do_them`, VerbNoun-naming, `call`.

Maybe these could be `SegmentedUsers::Upsert` and `SegmentedUsers::Delete` ?

Comment (filleduchaos) [reply app/services/bulk_segmented_users.rb:1]: Oh that's fair, and I don't think you should be reluctant about it! Ridhwana mentioned during onboarding but I forgot 😅

File: app/services/bulk_segmented_users.rb (hunk 1 above)
Comment (filleduchaos) [reply app/services/bulk_segmented_users.rb:25]: Ah - there is documentation for this class (especially for this decision) that didn't make it into this branch. I'll rectify that

**Comment by rt4914** [comment]
Un-assigning myself as I am not fully sure about the review.

File: app/services/segmented_users/bulk_upsert.rb [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      result
+    end
+
+    def perform_upsert(user_ids_batch)
+      segmented_users = build_records(user_ids_batch)
+
+      # We only want to touch the `updated_at` column of records that already exist,
+      # but specifying that causes ActiveRecord to emit malformed SQL (multiple assignments to the same column).
+      # Turning off Rails' automatic timestamp management via `record_timestamps`
+      # and managing timestamps ourselves yields the correct query.
```
Comment (jaw6) [inline app/services/segmented_users/bulk_upsert.rb:59]: 🎉

**Review by jaw6 — APPROVED** [review]
Two small questions, maybe these are ideas for future enhancement:

* It seems it's currently needing two API calls to `create` and later `add_users`, this might not be necessary, if `create` could also accept a list of `user_ids` ?
* It's possibly over-cautious, but there's currently no guards (particularly on `add_users` or `remove_users` that checks that these audience segments _must_ be `manual`-type segments — changing these segments would be temporary, at best, since they auto-refresh, but might be something to consider.

**Comment by filleduchaos** [comment]
@jaw6 

> It seems it's currently needing two API calls to `create` and later `add_users`, this might not be necessary, if `create` could also accept a list of `user_ids` ?

That's a good shout - I wasn't sure at the time what the UI/usage of these endpoints would look like. Maybe in a separate PR?

> * It's possibly over-cautious, but there's currently no guards (particularly on `add_users` or `remove_users` that checks that these audience segments _must_ be `manual`-type segments — changing these segments would be temporary, at best, since they auto-refresh, but might be something to consider.

Oh there is a [manual-only scope](https://github.com/forem/forem/blob/5a4075b76f60535d9304c4dcccaa50ddd134741d/app/controllers/api/v1/audience_segments_controller.rb#L53-L55) in the controller, as well as [a shared example](https://github.com/forem/forem/blob/5a4075b76f60535d9304c4dcccaa50ddd134741d/spec/requests/api/v1/audience_segments_spec.rb#L35-#40) in the request spec testing that they only operate on manual segments

**Pull request merged**